<img src="images/logodwengo.png" alt="Banner" width="150"/>

<div>
    <font color=#690027 markdown="1">
        <h1>IP-ADRES</h1> 
    </font>
</div>

<div class="alert alert-box alert-success">
    In deze notebook leer je wat een IP-adres is en wat binaire en hexadecimale getallen daarmee te maken hebben.
</div>

Het *internet* geeft toegang tot het world wide web, waarop men allerhande informatie kan vinden via websites. <br>
Elke website heeft een unieke locatie op het internet, weergegeven door zijn internetprotocol-adres, kortom *IP-adres*. (Een *protocol* is een set regels die vastleggen hoe twee digitale apparaten met elkaar kunnen communiceren).<br>
IP-adressen bestaan uit een rij cijfers (meestal vier getallen). De website van AI Op School bv. heeft $157.193.244.98$ als IP-adres. <br><br>
Zo'n IP-adres is niet zo gemakkelijk te onthouden, daarom gebruiken we *domeinnamen* zoals www.aiopschool.be. De omzetting van de domeinnaam naar het IP-adres gebeurt via het internetprotocol Domain Name System (DNS). <br>
Als een domeinnaam wordt ingegeven in de webbrowser, dan zal de browser de DNS-server contacteren. Op deze server staat een database van de domeinnamen met hun overeenkomstig IP-adres. De browser zal de DNS-server vragen naar het IP-adres van de gewenste website, waarna hij de opgevraagde website zal tonen. 
<img src="images/browser.png" alt="Banner" width="500"/>

<div>
    <font color=#690027 markdown="1">
        <h2>1. Het decimale en het binaire talstelsel</h2> 
    </font>
</div>

In het alledaagse leven gebruik je (natuurlijke) getallen uit het *tiendelige talstelsel*. Je werkt daarbij met eenheden, tientallen, hondertallen ... <br>
$10$ is de basis van het tiendelig talstelsel. <br> 
Een **cijfer** is een getal uit de verzameling $\{0,1,2,3,4,5,6,7,8,9\}$.<br>
Andere (natuurlijke) getallen worden gevormd door een **opeenvolging van deze cijfers**, zoals $23\,481$ en $5\,600\,934\,880$.<br>
Maar wat betekent dat nu concreet?<br><br>
Wel, het getal $532$ is opgebouwd uit $2$ eenheden, $3$ tientallen en $5$ hondertallen. 
<img src="images/tientallig.jpg" alt="Banner" width="230"/><br>
D.w.z. $532 = 5.100 + 3.10 + 2.1$ of m.a.w. $532 = 5.10^2 + 3.10^1 + 2.10^0$. <br>
Het aantal eenheden, tientallen, hondertallen ... worden dus gegeven door een cijfer. **Een getal in het tiendelig talstelsel wordt dus opgebouwd a.d.h.v. machten van $10$.** De plaats van een cijfer bepaalt hoeveel het bijdraagt aan het getal. <br> <br>
Je kan ook werken met een talstelsel met een andere basis, bv. $2$ voor het tweedelige of binaire talstelsel of $16$ voor het zestiendelige of hexadecimale talstelsel. Het binaire en hexadecimale talstelsel zijn veelgebruikt in de informatica.

In het *binaire talstelsel* beschik je slechts over twee cijfers, nl. $0$ en $1$. <br>
Andere binaire getallen worden dus gevormd door een opeenvolging van nullen en enen, zoals $1001001$ en $11111$. Ze zijn opgbebouwd a.d.h.v. machten van $2$, aangezien $2$ de basis van het binaire talstelsel is.
<img src="images/binair.jpg" alt="Banner" width="200"/><br>
Concreet is het getal $101 = 1.2^2 + 0.2^1 + 1.2^0$.<br>
D.w.z. dat het binaire getal $101$ overeenkomt met het getal $5$ in het tiendelige talstelsel.<br>
Dit binaire getal lees je als "één nul één". <br><br>

<div class="alert alert-block alert-info"> 
Een <b>bit</b> is een informatie-eenheid. De term is afkomstig van <b>binary digit</b>. Het is een eenheid die enkel de waarden 0 en 1 kan aannemen. <br>Acht bits vormen samen een <b>byte</b>. Een byte is is dus een opeenvolging van 8 nullen en enen. 
</div>

### Opdracht 1.1
- Hoeveel bytes kan je vormen? M.a.w. hoeveel binaire getallen kan je voorstellen met een byte?
- Welke tiendelige getallen kan je daarmee voorstellen?

Antwoord:

Antwoord: 
- Omdat $2^8 = 256$ kan men 256 verschillende getallen voorstellen met een byte. 
- Je kan er de natuurlijke getallen van $0$ t.e.m. $255$ mee voorstellen.

### Opdracht 1.2
Het IP-adres van de website van AI Op School is 157.193.244.98. Het bestaat uit vier tiendelige getallen.
- Hoe weet je dat de 4 getallen kunnen opgeslagen worden in een byte, dus 8 bits?
- Zet de vier getallen om naar hun binaire vorm.

Antwoord:

Antwoord:
- De 4 getallen liggen tussen 0 en 255, 0 en 255 inbegrepen, dus kunnen worden opgeslagen in een byte (zie opdracht 1.1).
- $157 = 1 + 156 = 1 + 128 + 28 = 1 + 128 + 16 + 12 = 128 + 16 + 8 + 4 + 1 =  2^0 + 2^2 + 2^3 + 2^4 + 2^7$  dus ........ dus $10011101$.
- $193 = 1 + 192 = 1 + 128 + 64 = 2^0 + 2^7 + 2^6$  dus ........ dus $11000001$.
- $244 = 128 + 116 = 128 + 64 + 52 = 128 + 64 + 32 + 20 = 128 + 64 + 32 + 16 + 4 = 2^7 +  2^6 + 2^5 + 2^4 + 2^2 $  dus ........ dus $11110100$.
- $98 = 64 + 34 = 64 + 32 + 2 = 2^6 + 2^5 + 2^1$  dus ....... dus $1100010$.

<div class="alert alert-box alert-info">
    Een IP-adres bestaat meestal uit vier tiendelige getallen die opgeslagen worden in 8 bits. <br>
    Een binair getal van 8 bits stelt een natuurlijk getal van 0 t.e.m. 255 (in het tiendelige talstelsel) voor.<br>
    Dus een IP-adres bestaat meestal uit vier natuurlijke getallen van 0 t.e.m. 255 (IPv4), of dus uit 4 binaire getallen van 8 cijfers.
</div>

<div> 
    <font color=#690027 markdown="1">
        <h2>2. Van het tiendelige talstelsel naar het binaire en omgekeerd</h2> 
    </font>
</div>

Het gebruikte talstelsel kan je meegeven door het toevoegen van een index. Voor het binaire getal $101$ schrijf je dan $101_{2}$ en voor het tiendelige getal $5$ schrijf je $5_{10}$.

### Voorbeeld

Je kan Python de *conversie* (omzetting) van binair naar tiendelig en omgekeerd laten maken. Test dit eens uit door de volgende code-cel uit te voeren. 

In [1]:
int("101", 2)        # binair getal omzetten naar zijn tiendelige gedaante

5

<div class="alert alert-box alert-info">
    De code <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">int("101", 2)</span> betekent dat het getal <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">101</span> uit het talstelsel met basis <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">2</span> wordt omgezet naar zijn voorstelling als een geheel getal, dus een getal uit het tiendelige talstelsel. <br>
    <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">int()</span> is een functie. Deze functie beschikt over twee parameters: een string die moet worden omgezet en de basis van het talstelsel die hoort bij het getal in de string. De waaarden van deze parameters, nl. <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">"101"</span> en <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">2</span>, worden argumenten genoemd.
</div>

Of omgekeerd kan je door de volgende code-cel uit te voeren, $5$ omzetten naar zijn binaire gedaante.

In [2]:
bin(5)                # tiendelig getal omzetten naar zijn binaire gedaante

'0b101'

<div class="alert alert-box alert-info">
    De code <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">bin(5)</span> betekent dat het getal <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">5</span> uit het tiendelige talstelsel wordt omgezet naar zijn voorstelling in het binaire talstelsel. De prefix <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">0b</span> benadrukt dat het een binair getal is.<br>
    <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">bin()</span> is een functie, die een integer (een geheel (tiendelig) getal) als parameter heeft, het getal dat omgezet moet worden. De waarde van deze parameter is hier <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">5</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">5</span> is het argument.
</div>

### Opdracht 2.1
- Zet $10101_{2}$ om naar zijn tiendelige vorm en omgekeerd. 
- Controleer met Python.

Antwoord:

Antwoord: $10101 =  1 . 2^0 + 0 . 2^1 + 1 . 2^2 + 0 . 2^3 + 1 . 2^4 = 1 + 4 +16 = 21$.<br>
Omgekeerd: $21 = 1 + 20 = 1 + 16 + 4 = 2^0 + 2^4 + 2^2$  dus ..... dus $10101$. 

In [3]:
# voorbeeldscript
int("10101", 2)

21

In [4]:
# voorbeeldscript
bin(21)

'0b10101'

### Opdracht 2.2
- Zet $3456_{10}$ om naar zijn binaire gedaante en omgekeerd. 
- Controleer met Python.

Antwoord:

Antwoord: $3456 = 2^11 + 1408 = 2^11 + 2^10 + 384 = 2^11 + 2^10 + 2^8 + 128 = 2^11 + 2^10 + 2^8 + 2^7$ <br>
dus ............ dus $110110000000$.<br>
Omgekeerd: $110110000000 = 1 . 2^11 + 1 . 2^10 + 0 . 2^9 + 1 . 2^8 + 1 . 2^7 = 4356$.<br> 

In [5]:
# voorbeeldscript
bin(3456)

'0b110110000000'

In [6]:
# voorbeeldscript
int("110110000000", 2)

3456

### Opdracht 2.3
- Controleer jouw resultaten bij opdracht 1.2 met Python.

In [7]:
# voorbeeldscript
bin(157)

'0b10011101'

In [8]:
# voorbeeldscript
bin(193)

'0b11000001'

In [9]:
# voorbeeldscript
bin(244)

'0b11110100'

In [10]:
# voorbeeldscript
bin(98)

'0b1100010'

<div>
    <font color=#690027 markdown="1">
        <h2>3. Het hexadecimale talstelsel</h2> 
    </font>
</div>

Omdat er een tekort was aan IP-adressen, werden IP-adressen van 8 hexadecimale getallen van 4 cijfers in het leven geroepen. Een IP-adres volgens deze nieuwe standaard, IPv6, ziet eruit als $0:0:0:0:0:\text{f}\, \text{f}\, \text{f}\, \text{f}\,:9 \, \text{d} \, \text{c} \, 1:\text{f} \, 4 \,6 \,2$ (AI Op School).

Om zo'n hexadecimaal getal van 4 cijfers op te slaan zijn 16 bits nodig (voor de getallen van $0$ t.e.m $15$ volstaan 4 bits). Je kan met de hexadecimale getallen van 4 cijfers de natuurlijke getallen van $0$ t.e.m. $65 \, 535$ voorstellen (van $0$ t.e.m $\text{f} \, \text{f} \, \text{f} \, \text{f}$).  

### Opdracht 3.1
- Geef het IP-adres http://[0:0:0:0:0:ffff:53d9:4b5b] in in je browser. Wat verschijnt er op je scherm? 

Antwoord:

Antwoord: It works!

In het *hexadecimale talstelsel* beschik je over 16 cijfers, eigenlijk de tiendelige getallen van $0$ t.e.m. $15$, maar voorgesteld door respectievelijk $ 0,1,2,3,4,5,6,7,8,9,a,b,c,d,e,f $. Hierbij staat $a$ voor $10$, $b$ voor $11$, ..., en $f$ voor $15$.

### Opdracht 3.2
Vul hieronder aan: <br><br>
Hexadecimale getallen worden dus gevormd door elementen uit ..... en zijn opgbebouwd a.d.h.v. ....<br>
<img src="images/hexadecimaal.jpg" alt="Banner" width="500"/><br>
Wat betekent dit concreet voor het hexadecimale getal $2 \, 3 \, \text{b} \, \text{e} \, 0 \, \text{d} \, \text{a} \, 2 \, \text{e}$? <br>
D.w.z. dat het hexadecimale getal $2 \, 3 \, \text{b} \, \text{e} \, 0 \, \text{d} \, \text{a} \, 2 \, \text{e}$ overeenkomt met het getal $....$ uit het tiendelige talstelsel.

Ingevuld:

Hexadecimale getallen worden dus gevormd door elementen uit $\{0, ..., 9, a, ..., f\}$ en zijn opgbebouwd a.d.h.v. machten van 16.<br>
<img src="images/hexadecimaal.jpg" alt="Banner" width="500"/><br>

D.w.z. dat het hexadecimale getal $2 \, 3 \, \text{b} \, \text{e} \, 0 \, \text{d} \, \text{a} \, 2 \, \text{e}$ overeenkomt met het getal $9 \; 594 \; 526 \; 254$ uit het tiendelige talstelsel.

<div>
    <font color=#690027 markdown="1">
        <h2>4. Van het tiendelige talstelsel naar het hexadecimale talstelsel en omgekeerd</h2> 
    </font>
</div>

Voer de volgende code-cellen uit.

In [11]:
hex(255)                        # hexadecimaal getal omzetten naar zijn tiendelige gedaante

'0xff'

In [12]:
int("ff", 16)

255

In [13]:
int("23be0da2e", 16)

9594526254

In [14]:
hex(9594526254)                 # tiendelig getal omzetten naar zijn hexadecimale gedaante

'0x23be0da2e'

In [15]:
int("23BE0DA2E", 16)

9594526254

<div class="alert alert-box alert-info">
    De prefix <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">0x</span> benadrukt dat het een hexadecimaal getal is.<br>
    <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">hex()</span> is een functie, die een integer (een geheel (tiendelig) getal) als parameter heeft, het getal dat omgezet moet worden. 
</div>

### Opdracht 4.1
- Hoeveel hexadecimale getallen kan je vormen met lengte 4? 
- Welke tiendelige getallen kan je daarmee voorstellen?

In [16]:
# voorbeeldscript
16**4

65536

In [17]:
# voorbeeldscript
# ffff (proef)
15 * 16**3 + 15 * 16**2 + 15 * 16**1 + 15 * 16**0

65535

Antwoord:

Antwoord:
- Omdat er voor elk cijfer 16 mogelijkheden zijn en het hexadecimale getal bestat uit 4 cijfers, en omdat $16^4 = 65 \, 536$ kan je $65 \, 536$ getallen vormen met lengte 4. 
- Je kan er de natuurlijke getallen van $0$ t.e.m. $65 \, 535$ mee voorstellen. 

### Opdracht 4.2
- Hoeveel hexadecimale getallen kan je vormen met lengte 8? 
- Welke tiendelige getallen kan je daarmee voorstellen?
- Hoeveel van deze getallen kan je voorstellen met een byte?
- Hoeveel bytes heb je nodig om al deze getallen te kunnen voorstellen?

In [18]:
# voorbeeldscript
16**8

4294967296

Antwoord:

Antwoord:
- Omdat er voor elk cijfer 16 mogelijkheden zijn en het hexadecimale getal opgebouwd is met 8 cijfers, en omdat $16^8 = 4 \, 294 \, 967 \, 296$ kan je ongeveer $4,3 . 10^9$ getallen vormen met lengte 8. 
- Je kan er de natuurlijke getallen van $0$ t.e.m. $4 \, 294 \, 967 \, 295$ mee voorstellen.
- Je kan er daarvan 256 voorstellen met een byte, nl. $0$ t.e.m. $255$.
- Merk op dat $16^8 = (2^8)^4$. Je hebt dus 4 bytes nodig.   

### Opdracht 4.3
- Hoeveel hexadecimale getallen kan je vormen met lengte 2? 
- Hoeveel binaire getallen kan je vormen met lengte 2?
- Hoeveel tiendelige getallen kan je vormen met lengte 2?

In [19]:
# voorbeeldscript
16**2

256

Antwoord:

Antwoord:
- Omdat er voor elk cijfer 16 mogelijkheden zijn en er twee cijfers gebruikt worden, en omdat $16^2 = 256$ kan je $256$ hexadecimale getallen vormen met lengte 2. Dat betekent dat je slechts 2 cijfers nodig hebt om de natuurlijke getallen van $0$ t.e.m $255$ hexadecimaal voor te stellen. 
- Omdat er voor elk cijfer 2 mogelijkheden zijn en er twee cijfers gebruikt worden, en omdat $2^2 = 4$ kan je $4$ binaire getallen vormen met lengte 2. 
- Omdat er voor elk cijfer 10 mogelijkheden zijn en er twee cijfers gebruikt worden, en omdat $10^2 = 100$ kan je $100$ tiendelige getallen vormen met lengte 2. 

<div class="alert alert-box alert-info">
    Als je met binaire getallen grote getallen wilt voorstellen, heb je al gauw heel wat bits nodig. Zo'n groot getal wordt dan al gauw moeilijk leesbaar. Hexadecimale getallen zijn daar meer geschikt voor. <br><br>
    Om de natuurlijke getallen van 0 t.e.m. 255 binair voor te stellen heb je 8 cijfers nodig, hexadecimaal zijn 2 cijfers daarvoor genoeg.
</div>

### Opdracht 4.4
In een computer kunnen kleuren in een *RGB-systeem* worden opgeslagen (RGB = rood, groen, blauw). De intensiteit van de aanwezige rode, groene en blauwe tinten worden elk weergegeven door een (tiendelig) getal van 0 t.e.m. 255. Wit bv. heeft als RGB(255, 255, 255) en zwart (0,0,0).<br> Binair heb je voor een RGB-code 24 cijfers nodig, 8 bits voor elk. Hexadecimaal kan dat al met 6 cijfers, bv. #ffffff voor wit, #000000 voor zwart en #0000ff voor blauw. 
- Bepaal de RGB-code voor blauw.
- 111111111010010100000000 is de binaire voorstelling van de RGB-code voor oranje. Bepaal de tiendelige voorstelling voor oranje.
- Bepaal de hexadecimale code voor oranje. 

In [20]:
# voorbeeldscript
int("ff", 16)

255

In [21]:
# voorbeeldscript
int("10100101", 2)

165

In [22]:
# voorbeeldscript
hex(165)

'0xa5'

Antwoord:

Antwoord:
- Omdat $00$ overeenkomt met 0 en $ff$ met 255, is de RGB-code van blauw (0, 0, 255). Dat is logisch: de intensiteiten van de rode en groene tinten is nul en de intensiteit van blauw de grootst mogelijke.
- Omdat $11111111$ overeenkommt met 255 en $00000000$ overeenkomt met 0 en berekend werd dat $10100101$ 165 is, is de RGB-code voor oranje (255, 165, 0).
- De omzetting van 255, 165 en 0 naar hexadecimale vorm geeft $ff$, $a5$ en $00$, dus de hexadecimale code voor oranje is #ffa500. 

<div class="alert alert-box alert-warning">
    Meer over RGB leer je in de notebook 'Tensoren en RGB' over digitale beeldverwerking. 
</div>

<div>
    <font color=#690027 markdown="1">
        <h2>5. Oefeningen</h2> 
    </font>
</div>

### Opdracht 5.1
Welke van de volgende getallen heeft in zijn binaire gedaante het meeste bits die 1 zijn?
- Het tiendelige getal 100;
- het binaire getal 1010;
- het tiendelige getal 79;
- het hexadecimale getal c5;
- het hexadecimale getal a2.

In [23]:
# voorbeeldscript
bin(100)

'0b1100100'

In [24]:
# voorbeeldscript
bin(79)

'0b1001111'

In [25]:
# voorbeeldscript
bin(int("c5", 16))

'0b11000101'

In [26]:
# voorbeeldscript
bin(int("a2", 16))

'0b10100010'

Antwoord:

Antwoord: Het tiendelige getal 79 heeft het meest enen in zijn binaire gedaante, nl. 5.

### Opdracht 5.2
Welke van de volgende getallen is het grootst?
- het hexadecimale getal 1b9
- het binaire getal 1001000100
- het hexadecimale getal 5c2
- het tiendelige getal 670

In [27]:
# voorbeeldscript
int("1b9", 16)

441

In [28]:
# voorbeeldscript
int("1001000100", 2)

580

In [29]:
# voorbeeldscript
int("5c2", 16)

1474

Antwoord:

Antwoord: Het hexadecimale getal 5c2 is het grootst.

### Opdracht 5.3
Om een kleur weer te geven kan je RGB-code of een hexadecimaal getal gebruiken.<br>
Welke kleur stelt het binaire getal 101010101111000011010001 voor? 

In [30]:
# voorbeeldscript
hex(int("101010101111000011010001", 2))

'0xaaf0d1'

Antwoord:

Antwoord: De kleur is 'magic mint'.

### Opdracht 5.4
- Hoeveel IP-adressen kan je vormen met de standaard IPv4 (de 4 natuurlijke getallen)? 
- Hoeveel met de standaard IPv6 (de 8 hexadecimale getallen)?
- Wat is het verschil in grootte-orde?

In [31]:
# voorbeeldscript
256**4

4294967296

In [32]:
# voorbeeldscript
16**32

340282366920938463463374607431768211456

Antwoord:

Antwoord:
-  Een IPv4-adres bestaat uit 4 natuurlijke getallen van $0$ t.e.m. $255$. Omdat $256^4 = 4 \, 294 \, 967\, 296$, kan je ongeveer $4,3 . 10^9$ zulke IP-adressen vormen. 
-  Een IPv6-adres bestaat uit 8 getallen van $0$ t.e.m. $16^4-1$. Voor elk van de getallen zijn er dus $16^4$ mogelijkheden. Omdat $(16^4)^8 = 340 \, 282 \, 366 \, 920 \, 938 \, 463 \, 463 \, 374 \, 607 \, 431 \, 768 \, 211 \, 456$, dus je kan ongeveer $3,4 . 10^{38}$ zulke IP-adressen vormen. 
-  Het verschil in grootte-orde is $10^{29}$.

### Opdracht 5.5
Zet de getallen in het IP-adres $0:0:0:0:0:\text{f}\, \text{f}\, \text{f}\, \text{f}\,:9 \, \text{d} \, \text{c} \, 1:\text{f} \, 4 \,6 \,2$ van AI Op School om naar tiendelige getallen.

In [33]:
# voorbeeldscript
int("ffff", 16)

65535

In [34]:
# voorbeeldscript
int("9dc1", 16)

40385

In [35]:
# voorbeeldscript
int("f462", 16)

62562

Antwoord:

Antwoord:
- De nullen zijn ook nul als tiendelig getal.
- Het IP-adres is dus: $0 \, : \, 0 \, : \, 0 \, : \, 0 \, : 0 \, : 65535 \, : \,  40385 \, : \,  62562$.

<div>
    <font color=#690027 markdown="1">
        <h2>6. Onthoud</h2> 
    </font>
</div>

<div class="alert alert-block alert-info"> 
    Elke website heeft een unieke locatie op het internet, weergegeven door zijn IP-adres.<br>
IP-adressen bestaan uit een rij cijfers:<br>
- meestal vier getallen, elk een tiendelig getal opgeslagen in 8 bits (IPv4); <br>
- 8 hexadecimale getallen, elk opgeslagen in 16 bits, bij de nieuwste IP-adressen (IPv6). 

De IPv4-adressen gebruiken 32 bits, de IPv6-adressen gebruiken er 128. <br>
Het aantal mogelijke  IP-adressen is bij de nieuwe standaard aanzienlijk groter dan bij de oude (verschil van $10^{29}$ in grootte-orde ).
</div>

<div>
    <h2>Kijktip</h2> 
</div>

Rob Hoevenaars. 1 juni 2016. *Talstelsels: uitleg en rekenen (binair, decimaal ...).* https://youtu.be/eCqtuKQX_ac <br>
Rob Hoevenaars. 1 juni 2016. *Talstelsels hexadecimaal: uitleg!* https://youtu.be/GFi40c4KRY8

<div>
    <h2>Referentielijst</h2> 
</div>

Isaac Computer Science. Network fundamentals. https://isaaccomputerscience.org/topics/networking
Isaac Computer Science. DNS and the cloud. https://isaaccomputerscience.org/concepts/gcse_net_dns
Isaac Computer Science. Secondary storage. 
     https://isaaccomputerscience.org/concepts/sys_hard_secondary_storage
Isaac Computer Science. Client-server technologies. 
     https://isaaccomputerscience.org/concepts/net_internet_client_server_model

<img src="images/cclic.png" alt="Banner" align="left" width="100"/><br><br>
Notebook Python in wiskunde, zie <a href="http://www.aiopschool.be">AI Op School</a>, van F. wyffels & N. Gesquière is in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>.